aea/helpers/async_utils.py

```diff
@@ -0,0 +1,324 @@
+# -*- coding: utf-8 -*-
+# ------------------------------------------------------------------------------
+#
+#   Copyright 2018-2019 Fetch.AI Limited
+#
+#   Licensed under the Apache License, Version 2.0 (the "License");
+#   you may not use this file except in compliance with the License.
+#   You may obtain a copy of the License at
+#
+#       http://www.apache.org/licenses/LICENSE-2.0
+#
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+#
+# ------------------------------------------------------------------------------
+"""This module contains the misc utils for async code."""
+import asyncio
+import datetime
+import logging
+import time
+from asyncio import CancelledError
+from asyncio.events import AbstractEventLoop, TimerHandle
+from asyncio.futures import Future
+from asyncio.tasks import ALL_COMPLETED, FIRST_COMPLETED, Task
+from threading import Thread
+from typing import (
+    Any,
+    Callable,
+    Coroutine,
+    List,
+    Optional,
+    Sequence,
+    Set,
+    Tuple,
+    Union,
+    cast,
+)
+
+try:
+    from asyncio import create_task
+except ImportError:  # pragma: no cover
+    # for python3.6!
+    from asyncio import ensure_future as create_task  # type: ignore # noqa: F401
+
+
+logger = logging.getLogger(__file__)
+
+
+def ensure_list(value: Any) -> List:
+    """Return [value] or list(value) if value is a sequence."""
+    if not isinstance(value, (list, tuple)):
+        value = [value]
+    return list(value)
+
+
+class AsyncState:
+    """Awaitable state."""
+
+    def __init__(self, initial_state: Any = None, loop: AbstractEventLoop = None):
+        """Init async state.
+
+        :param initial_state: state to set on start.
+        :param loop: optional asyncio event loop.
+        """
+        self._state = initial_state
```

Comment:
Actually, if the above is true, this could be:
```python
self._state = AgentLoopStates(initial_state)
```

That is:
- if `initial_state` is an instance of `AgentLoopStates`, then nothing bad happens: it just returns the same enum value.
- if `initial_state` is a valid enum value (that is, the values defined on the right of the enumeration names, e.g. `None`, `"started"`, `"stopped"`, etc.) then the return value is the corresponding `AgentLoopStates` instance.